src/main/java/com/shapesecurity/functional/data/NonEmptyImmutableList.java

```diff
@@ -35,6 +37,7 @@ protected NonEmptyImmutableList(@Nonnull T head, @Nonnull final ImmutableList<T>
         super(tail.length + 1);
         this.head = head;
         this.tail = tail;
+        this.hashCode = HashCodeBuilder.put(tail.hashCode(), head);
```

Comment:
Can we compute this lazily only once `hashCode` is called, then cache it in an instance field?

Reply:
Yes, we could do a lazy iterative cache within hashCode. The main issue is that it prevents us from caching sublists without recalculation the hash due to the iteration order -- or using a stack beforehand. There is less overhead to this approach, however, if we don't call hashCode often, it might be appropriate to do it lazily instead.

Reply:
@michaelficarra The PR has been updated to be lazy.